src/auth.rs

```diff
@@ -3,41 +3,95 @@ use crate::log::Log;
 use crate::verify::verify;
 use anyhow::{anyhow, Context, Result};
 use ssh_agent_client_rs::Error as SACError;
-use ssh_key::public::KeyData;
-use ssh_key::{AuthorizedKeys, PublicKey};
+use ssh_agent_client_rs::Identity;
+use ssh_key::AuthorizedKeys;
+use ssh_key::PublicKey;
 use std::collections::HashSet;
 use std::path::Path;
 
 const CHALLENGE_SIZE: usize = 32;
 
-/// Finds the first key, if any, that the ssh-agent knows about that is also present
+/// Finds the first key or, if any, that the ssh-agent knows about that is also present
 /// in the file referenced by keys_file_path, sends a random message to be signed and
 /// verifies the signature with the public key.
+/// By passing a ca_keys_file_path, the function will also check if the agent has SSH
+/// certificates and if the certificate is signed by a trusted CA key present and
+/// perform the same signature verification process in addition to certificate
+/// validation checks.
 ///
 /// Returns Ok(true) if a key was found and the signature was correct, Ok(false) if no
 /// key was found, and Err if agent communication or signature verification failed.
 pub fn authenticate(
     keys_file_path: &str,
+    ca_keys_file_path: Option<&str>,
     mut agent: impl SSHAgent,
     log: &mut impl Log,
 ) -> Result<bool> {
     let keys = keys_from_file(Path::new(keys_file_path))?;
+    // If no CA keys file is provided, then initialize ca_keys as an empty set.
```

Comment:
This whole piece of functionality seems to me like something that would be well suited for moving into a separate function, with some unit tests.